### PR TITLE
Use ovsdb

### PR DESCRIPTION
**Pull request by SchSeba**
This PR changes the way the ovs-cni interact with the ovs.

With the PR we communicate directly via the ovsdb socket file.

This PR will the problem we have when we copy the binaries into the host.

### CODE REVIEW

File: pkg/ovsdb/ovsdb.go [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Not in use
+func (self *OvsDriver) CreateBridge(bridgeName string, failMode string, stp bool) error {
+	namedUuidStr := "dummy"
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: what is this?

File: pkg/ovsdb/ovsdb.go [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+		}
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Not in use
+func (self *OvsDriver) CreateBridge(bridgeName string, failMode string, stp bool) error {
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: We should never touch bridges, we just merely attach ports to existing ones. By that we make sure that the bridge is always created and configured by administrator as part of host networking setup.

File: pkg/ovsdb/ovsdb.go [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{condition},
+	}
+
+	operations := []libovsdb.Operation{brOp, mutateOp}
+	return self.ovsdbTransact(operations)
+}
+
+// Create an internal port in OVS
+func (self *OvsDriver) CreatePort(intfName, intfType string, vlanTag uint, contIfaceName, contNetnsPath, ovnPortName string) error {
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Could you make it so we pass contIfaceName, contNetnsPath and ovnPortName as a arbitrary map? This is hard to read when it is called and is too specific.

File: pkg/ovsdb/ovsdb.go [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+		return "", false, fmt.Errorf("error get more than one port")
+	}
+
+	return fmt.Sprintf("%v", operationResult.Rows[0]["name"]), true, nil
+
+}
+
+// ************************ Notification handler for OVS DB changes ****************
+func (self *OvsDriver) Update(context interface{}, tableUpdates libovsdb.TableUpdates) {
+	// fmt.Printf("Received OVS update: %+v\n\n", tableUpdates)
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Please drop code that is not being used.

File: pkg/ovsdb/ovsdb.go (hunk 2 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: remove this function

File: pkg/ovsdb/ovsdb.go [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+	if !ovsDriver.IsBridgePresent(bridgeName) {
+		return nil, fmt.Errorf("failed to find bridge %s", bridgeName)
+	}
+
+	// Return the new OVS driver
+	return ovsDriver, nil
+}
+
+// Populate local cache of ovs state
+func (self *OvsDriver) populateCache(updates libovsdb.TableUpdates) {
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: I remove the cache we don't need it.

we are using this package on the ovs-cni plugin that runs on every call we don't need to save cache here

File: pkg/ovsdb/ovsdb.go [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+		ovs.Register(ovsDriver)
+
+		// Populate initial state into cache
+		initial, _ := ovs.MonitorAll("Open_vSwitch", "")
+		ovsDriver.populateCache(*initial)
+	}()
+
+	// HACK: sleep the main thread so that Cache can be populated
+	// Add a retry loop here and not wait 1 second
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: remove it we don't need the cache to check if a bridge exist just make a select transaction.

**Comment by SchSeba** [comment]
@phoracek can you please take another look?

File: pkg/ovsdb/ovsdb.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
+	portUuidStr := infoMap["intfName"]
+	intfUuidStr := fmt.Sprintf("Intf%s", infoMap["intfName"])
+	portUuid := []libovsdb.UUID{{GoUUID: portUuidStr}}
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Oh, I thought we pass these values to OVS as plain strings. Now I see we have to convert them to specific format.

Sorry for the confusion. Would you mind then creating infoMap structure with specific types, or just revert the original code with all parameters listed...

File: pkg/ovsdb/ovsdb.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+	intfUuid := []libovsdb.UUID{{GoUUID: intfUuidStr}}
+	opStr := "insert"
+	var err error = nil
+
+	// insert/delete a row in Interface table
+	intf := make(map[string]interface{})
+	intf["name"] = infoMap["intfName"]
+	//intf["type"] = ""
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: please drop unused code

File: pkg/ovsdb/ovsdb.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
+	portUuidStr := infoMap["intfName"]
+	intfUuidStr := fmt.Sprintf("Intf%s", infoMap["intfName"])
+	portUuid := []libovsdb.UUID{{GoUUID: portUuidStr}}
+	intfUuid := []libovsdb.UUID{{GoUUID: intfUuidStr}}
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Why do we convert to UUID anyway?

File: pkg/ovsdb/ovsdb.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
+	portUuidStr := infoMap["intfName"]
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: could you please reorganize the code a little?

```
portUuidStr := ...
portUuid := ...

intf...
...

opStr :=

```

File: pkg/ovsdb/ovsdb.go [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+	intfUuidStr := fmt.Sprintf("Intf%s", infoMap["intfName"])
+	portUuid := []libovsdb.UUID{{GoUUID: portUuidStr}}
+	intfUuid := []libovsdb.UUID{{GoUUID: intfUuidStr}}
+	opStr := "insert"
+	var err error = nil
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: why do we need this?

File: pkg/ovsdb/ovsdb.go [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+	// insert/delete a row in Interface table
+	intf := make(map[string]interface{})
+	intf["name"] = infoMap["intfName"]
+	//intf["type"] = ""
+
+	if infoMap["ovnPortName"] != "" {
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Please document what is going on here

File: pkg/ovsdb/ovsdb.go [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+		}
+	}
+
+	// Return success
+	return nil
+}
+
+// **************** OVS driver API ********************
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: This function is too long and contains too much logic. Could you please try to reorganize it a little? Add more inline documentation on Ports and Interfaces.

File: pkg/ovsdb/ovsdb.go [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+	operationResult := transactionResult[0]
+	if operationResult.Error != "" {
+		return fmt.Errorf("%s - %s", operationResult.Error, operationResult.Details)
+	}
+
+	if len(operationResult.Rows) != 1 {
+		return fmt.Errorf("failed to find port")
+	}
+
+	portUuidStr := operationResult.Rows[0]["_uuid"].([]interface{})
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Please comment on what are you doing here

File: pkg/ovsdb/ovsdb.go [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+	portUuid := []libovsdb.UUID{{GoUUID: fmt.Sprintf("%v", portUuidStr[1])}}
+	opStr := "delete"
+
+	// insert/delete a row in Interface table
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: Insert/delete? Aren't you always deleting?

File: pkg/ovsdb/ovsdb.go [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+	if len(operationResult.Rows) != 1 {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// Return ovs port name for an container interface
+func (self *OvsBridgeDriver) GetOvsPortForContIface(contIface, contNetnsPath string) (string, bool, error) {
+	seachMap := map[string]string{"contNetns": contNetnsPath, "contIface": contIface}
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: search?

File: pkg/ovsdb/ovsdb.go [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+	if operationResult.Error != "" {
+		return "", false, fmt.Errorf("%s - %s", operationResult.Error, operationResult.Details)
+	}
+
+	if len(operationResult.Rows) != 1 {
+		return "", false, fmt.Errorf("error get more than one port")
+	}
+
+	return fmt.Sprintf("%v", operationResult.Rows[0]["name"]), true, nil
+
```
Comment (alonSadan) [inline pkg/ovsdb/ovsdb.go]: maybe remove empty line here

File: pkg/ovsdb/ovsdb.go (hunk 9 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: for the json convert `GoUUID string "json:\"uuid\""`

Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: for the json
```golang
// UUID is a UUID according to RFC7047
type UUID struct {
	GoUUID string `json:"uuid"`
}
```

File: pkg/ovsdb/ovsdb.go (hunk 11 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: I move it a bit more down in the code

File: pkg/ovsdb/ovsdb.go (hunk 8 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: done

File: pkg/ovsdb/ovsdb.go (hunk 12 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: done

File: pkg/ovsdb/ovsdb.go (hunk 17 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: ++

File: pkg/ovsdb/ovsdb.go (hunk 16 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: ++

File: pkg/ovsdb/ovsdb.go (hunk 15 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: removed

File: pkg/ovsdb/ovsdb.go (hunk 14 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: done

File: pkg/ovsdb/ovsdb.go [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+	if err != nil {
+		return err
+	}
+
+	portUuid, portOp, err := createPortOperation(intfName, contNetnsPath, contIfaceName, vlanTag, intfUuid)
+	if err != nil {
+		return err
+	}
+
+	// mutate the Ports column of the row in the Bridge table
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: the previous two helpers are nice! could you change this last bit to attachPortOperation?

File: pkg/ovsdb/ovsdb.go (hunk 15 above)
Comment (phoracek) [reply pkg/ovsdb/ovsdb.go]: My diff still shows insert/delete

File: pkg/ovsdb/ovsdb.go [hunk 19]
```diff
@@ -0,0 +1,329 @@
...
+		Where: []interface{}{condition},
+	}
+
+	transactionResult, err := self.ovsClient.Transact("Open_vSwitch", selectOp)
+	if err != nil {
+		return err
+	}
+
+	if len(transactionResult) != 1 {
+		return fmt.Errorf("unknow error")
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: unknown

File: pkg/ovsdb/ovsdb.go [hunk 20]
```diff
@@ -0,0 +1,329 @@
...
+	defer self.lock.RUnlock()
+
+	condition := libovsdb.NewCondition("name", "==", intfName)
+	selectOp := libovsdb.Operation{
+		Op:    "select",
+		Table: "Port",
+		Where: []interface{}{condition},
+	}
+
+	transactionResult, err := self.ovsClient.Transact("Open_vSwitch", selectOp)
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: you already have a helper function for this operation

File: pkg/ovsdb/ovsdb.go [hunk 21]
```diff
@@ -0,0 +1,329 @@
...
+
+	condition := libovsdb.NewCondition("external_ids", "==", ovsmap)
+	selectOp := libovsdb.Operation{
+		Op:      "select",
+		Table:   "Port",
+		Where:   []interface{}{condition},
+		Columns: []string{"name", "external_ids"},
+	}
+
+	transactionResult, err := self.ovsClient.Transact("Open_vSwitch", selectOp)
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: can you use transaction helper you defined above?

File: pkg/ovsdb/ovsdb.go [hunk 22]
```diff
@@ -0,0 +1,329 @@
...
+}
+func (self *OvsBridgeDriver) Locked([]interface{}) {
+}
+func (self *OvsBridgeDriver) Stolen([]interface{}) {
+}
+func (self *OvsBridgeDriver) Echo([]interface{}) {
+}
+
+// ************************ Helper functions ********************
+func createInterfaceOperation(intfName, ovnPortName string) ([]libovsdb.UUID, *libovsdb.Operation, error) {
```
Comment (phoracek) [inline pkg/ovsdb/ovsdb.go]: can we make the logic of this module simpler if we always use the same name for port and interface?

File: pkg/plugin/plugin.go [hunk 23]
```diff
@@ -189,24 +191,15 @@ func getBridgeName(bridgeName, ovnPort string) (string, error) {
...
-func attachIfaceToBridge(hostIfaceName string, contIfaceName string, brName string, vlanTag *uint, contNetnsPath string, ovnPortName string) error {
-	// Set external IDs so we are able to find and remove the port from OVS
-	// database when CNI DEL is called.
-	command := []string{
-		"--", "add-port", brName, hostIfaceName,
-		"--", "set", "Port", hostIfaceName, fmt.Sprintf("external-ids:contNetns=%s", contNetnsPath),
-		"--", "set", "Port", hostIfaceName, fmt.Sprintf("external-ids:contIface=%s", contIfaceName),
-	}
+func attachIfaceToBridge(ovsDriver *ovsdb.OvsBridgeDriver, hostIfaceName string, contIfaceName string, vlanTag *uint, contNetnsPath string, ovnPortName string) error {
+	var vlanTagNum uint = 0
```
Comment (phoracek) [inline pkg/plugin/plugin.go]: Let's make it `int` and use -1 to enable trunk. I vaguely remember VLAN 0 had some special significance.

File: pkg/ovsdb/ovsdb.go (hunk 22 above)
Comment (SchSeba) [reply pkg/ovsdb/ovsdb.go]: the OVN port name is different from the veth name.

File: pkg/plugin/plugin.go (hunk 23 above)
Comment (SchSeba) [reply pkg/plugin/plugin.go]: I think this change need to go to another PR is not related to this one.

we need to change the netConf struct
```
type netConf struct {
	types.NetConf
	BrName  string `json:"bridge,omitempty"`
	VlanTag *uint  `json:"vlan"`
}
```

and also this mean we need to update the documents and tell the user to config -1 if he want to use trunk (not so nice in my opinion)

Comment (phoracek) [reply pkg/plugin/plugin.go]: Ah, darn, sure, get it now.

**Pull request merged**